provider/apple.go

```diff
@@ -0,0 +1,532 @@
+	data.Set("code", code)
+	data.Set("redirect_uri", redirectURI) // redirect URL can't refer to localhost and must have trusted certificate and https protocol
+	data.Set("grant_type", "authorization_code")
+
+	client := http.Client{Timeout: time.Second * 5}
+	req, err := http.NewRequestWithContext(ctx, "POST", ah.endpoint.TokenURL, strings.NewReader(data.Encode()))
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("content-type", appleRequestContentType)
+	req.Header.Add("accept", AcceptHeader)
+	req.Header.Add("user-agent", ah.conf.UserAgent) // apple requires a user agent
+
+	res, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	// Apple REST API response either 200 (ok OK) or 400 (if error)
+	if res.StatusCode >= 400 {
+		return fmt.Errorf("exchange response error code: %d", res.StatusCode)
+	}
+
+	err = json.NewDecoder(res.Body).Decode(result)
+	defer func() {
+		if err = res.Body.Close(); err != nil {
+			ah.L.Logf("[ERROR] close request body failed when get access token: %v", err)
+		}
+	}()
+	return err
+}
+
+// getTokenClaims decodes the id_token in response for extract claims with user data
+func (ah AppleHandler) getTokenClaims(idToken string) (jwt.MapClaims, error) {
+	tkn, _, err := new(jwt.Parser).ParseUnverified(idToken, jwt.MapClaims{})
+
+	if err != nil {
+		return nil, err
+	}
+
+	claims, ok := tkn.Claims.(jwt.MapClaims)
+	if !ok {
+		return nil, errors.New("can't convert token claims to standard claims")
+	}
+	return claims, nil
+}
+
+// createClientSecret use for create the JWT client secret required to make requests to the Apple validation server.
+// for more details go to link: https://developer.apple.com/documentation/sign_in_with_apple/generate_and_validate_tokens#3262048
+func (ah *AppleHandler) createClientSecret() (string, error) {
+	if ah.conf.privateKey == nil {
+		return "", errors.New("private key can't be empty")
+	}
+	// Create the Claims
+	now := time.Now()
+	exp := now.Add(time.Second * 86400).Unix() // default value
+
+	if ah.conf.ClientSecretExpire != 0 {
+		exp = now.Add(time.Second * time.Duration(ah.conf.ClientSecretExpire)).Unix()
+	}
+
+	claims := &jwt.StandardClaims{
+		Issuer:    ah.conf.TeamID,
+		IssuedAt:  now.Unix(),
+		ExpiresAt: exp,
+		Audience:  "https://appleid.apple.com",
+		Subject:   ah.conf.ClientID,
+	}
+
+	tkn := jwt.NewWithClaims(jwt.SigningMethodES256, claims)
+	tkn.Header["alg"] = "ES256"
+	tkn.Header["kid"] = ah.conf.KeyID
+
+	return tkn.SignedString(ah.conf.privateKey)
+}
+
+func (ah *AppleHandler) isClientSecretExpired() (bool, error) {
+
+	tkn, _, err := new(jwt.Parser).ParseUnverified(ah.conf.clientSecret, jwt.MapClaims{})
+
+	if err != nil {
+		return false, err
+	}
+	claims, ok := tkn.Claims.(jwt.MapClaims)
+	if !ok {
+		return false, errors.New("can't convert token claims to standard claims")
+	}
+	return claims.VerifyExpiresAt(time.Now().Unix(), true), nil
+}
+func (ah *AppleHandler) parseUserData(user *token.User, jUser string) {
```

Comment:
Please make this function covered with tests, called from the exposed function if possible.